From 867ca0138fefc077d418ca3bb49793ae0d9d73ef Mon Sep 17 00:00:00 2001
From: Sebastian Fricke <sebastian.fricke.linux@gmail.com>
Date: Tue, 1 Dec 2020 19:17:03 +0100
Subject: [PATCH v2] Fix for NanoPC-T4: resolve regulator before creation

This patch reverts https://github.com/torvalds/linux/commit/aea6cb99703e17019e025aa71643b4d3e0a24413
that got backported into both 5.8.17 and 5.9.2

Without it rk808 used by all rk3399 boards fails to be probed.

Ported patch from armbian rockchip64-dev to 5.10-rc6.

Signed-off-by: Sebastian Fricke <sebastian.fricke.linux@gmail.com>
---
 drivers/regulator/core.c | 21 ++++++++-------------
 1 file changed, 8 insertions(+), 13 deletions(-)

diff --git a/drivers/regulator/core.c b/drivers/regulator/core.c
index 42bbd99a36ac..82acdee94187 100644
--- a/drivers/regulator/core.c
+++ b/drivers/regulator/core.c
@@ -5298,20 +5298,15 @@ regulator_register(const struct regulator_desc *regulator_desc,
 	else if (regulator_desc->supply_name)
 		rdev->supply_name = regulator_desc->supply_name;
 
+	/*
+	 * Attempt to resolve the regulator supply, if specified,
+	 * but don't return an error if we fail because we will try
+	 * to resolve it again later as more regulators are added.
+	 */
+	if (regulator_resolve_supply(rdev))
+		rdev_dbg(rdev, "unable to resolve supply\n");
+
 	ret = set_machine_constraints(rdev);
-	if (ret == -EPROBE_DEFER) {
-		/* Regulator might be in bypass mode and so needs its supply
-		 * to set the constraints */
-		/* FIXME: this currently triggers a chicken-and-egg problem
-		 * when creating -SUPPLY symlink in sysfs to a regulator
-		 * that is just being created */
-		ret = regulator_resolve_supply(rdev);
-		if (!ret)
-			ret = set_machine_constraints(rdev);
-		else
-			rdev_dbg(rdev, "unable to resolve supply early: %pe\n",
-				 ERR_PTR(ret));
-	}
 	if (ret < 0)
 		goto wash;
 
-- 
2.20.1

